### ISSUE TITLE
Issues packaging appJar - unable to import tkinter

### ISSUE THREAD

**Issue by tengerdata** [jarvisteach/appJar#122]
First, I have to say appJar is awesome! I am very happy with it. I've tried several GUIs and for me appJar is, by far, the best one! Thanks for doing this.

My issue:

I am on Python 3.5 and I pip installed tkinter. When I use appJar it works perfectly and I can also import and use tkinter with no problem. So, tkinter is installed and it works. And, appJar works as well so appjar.py has no problem importing tkinter.

The problem starts when I build an executable. The executable is built, but when I try to open the exe file I get an error that appJar can't find the tkinter module. "_No module named tkinter_" or something like that.

I am working in a conda environment but when I check the site-packages in the environment, there is no tkinter directory there. That must me because I didn't use conda install for tkinter, I used pip if I remember correctly. The reason I am not using conda install because for some reason I run into problems when I try. Btw I only have problems with tkinter install on condas, for everything else I use conda install and it works fine.

To summarize:
1. appjar.py has no problem importing tkinter in the development environment
2. appjar.py can not import tkinter when I buil an executable 

Any suggestions on how I can solve this issue?

Thanks.

**Comment by jacobthetechy** [comment]
appjar is awesome, and the problem your having is not with appjar its with the package you use to build an executable.

The fix is to include an import to tkinter in your app's code. such as:
```
#myapp.py
import tkinter
#your other imports

#your code.......
```

The reason:
I'm not sure which pachage your using but I have used pyinstaller for my projects. To my understanding the problem that it has and all the others have is nested/hidden imports. It will see that you are import appjar when going thought the compiling process but it will not then check every file in appjar to see if there are any nested/hidden imports that it need to also get. So when you run the exe it throws that error because it never included tkinter in the build.

**Comment by tengerdata** [comment]
@jacobthetechy Thanks for your reply. I just tried what you suggested, but unfortunately I still get the error: 
ImportError: No module named 'tkinter'

**Comment by jacobthetechy** [comment]
OK lets try this.

from appJar import gui
import tkinter as tk

I can't remember if the order was important or not, but this is what I have for my projects. Tho I am using 2.7 and tkinter in the import is Tkinter not tkinter.

**Comment by tengerdata** [comment]
@jacobthetechy It doesn't work. I tried, just about everything. I am kind of giving up on this.

**Comment by jacobthetechy** [comment]
What package are you using to make the executable?

**Comment by jarvisteach** [comment]
Sorry all, this is something I have zero experience of.

I've already raised an issue #117 to investigate the best way to package appJar into an executable, but won't get a chance to investigate until the Easter holidays.

A few questions:
* What are you using to create the package - is it pyinstaller?
* What platform are you on - it is Windows?
* Is appJar installed through pip?

When I first had a quick look into this, as jacob mentioned, the package builder doesn't necessarily search through all the code, to find out what libraries to include in the package. So, you may need to edit the config file, for the package builder, to tell it to include the tkinter library...

**Comment by jarvisteach** [comment]
So, I've had a quick play on the RaspberryPi - and it worked quite well...

If appJar is installed via pip, then the command:  
```pyinstaller -F demo.py```  
generated a working executable  

If appJar wasn't installed via pip, then I had to add the ```-p``` flag, to set the PYTHONPATH for appJar.

Looking through the forums, it seems there can be issues with tkinter imports, but according to (the docs)[https://pyinstaller.readthedocs.io/en/stable/usage.html#what-to-bundle-where-to-search] you can set a *hidden import* with the ```--hiddenimport tkinter`` flag

**Comment by jarvisteach** [comment]
I'm going to close this issue. It's not a problem with appJar, but rather the packaging tool being used. When everything is packaged up, you need to make sure tkinter is being included in the package.

There is a separate issue #117, already raised, to document best practices for packaging appJar on each platform - I've had success on both Linux & Windows, with the default installations. No luck yet on Mac OSX, but then my installation on there is a mess anyway...
